### PR TITLE
mgr/cephadm: check spec host when adding osd

### PR DESCRIPTION
**Pull request by rkachach**
Fixes: https://tracker.ceph.com/issues/47872

Signed-off-by: Redouane Kachach <rkachach@redhat.com>

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [x] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

File: src/pybind/mgr/orchestrator/module.py [hunk 1]
```diff
@@ -791,6 +791,15 @@ def _daemon_add_osd(self,
             return HandleCommandResult(-errno.EINVAL, stderr=usage)
         try:
             host_name, block_device = svc_arg.split(":")
+
+            hosts_completion = self.get_hosts()
+            hosts = raise_if_exception(hosts_completion)
+            filter_spec = PlacementSpec(host_pattern=host_name)
+            filtered_hosts: List[str] = filter_spec.filter_matching_hostspecs(hosts)
```
[the remark this answers is not in the document]
Comment (rkachach) [reply src/pybind/mgr/orchestrator/module.py]: I moved the code to the cephadm module (though the error message is more generic)

**Comment by rkachach** [comment]
jenkins retest this please

**Comment by rkachach** [comment]
@adk3798 UT has been added to check the new validation

**Review by adk3798 — APPROVED** [review]

**Comment by adk3798** [comment]
http://pulpito.front.sepia.ceph.com/adking-2022-03-23_02:54:35-orch:cephadm-wip-adk-testing-2022-03-22-2000-distro-basic-smithi/

2 Failures caused by wrong error code from host add command due to another PR included in the run

File: src/pybind/mgr/cephadm/tests/test_cephadm.py [hunk 2]
```diff
@@ -811,6 +811,12 @@ def test_create_osds(self, cephadm_module):
             c = cephadm_module.create_osds(dg)
             out = wait(cephadm_module, c)
             assert out == "Created no osd(s) on host test; already created?"
+            bad_dg = DriveGroupSpec(placement=PlacementSpec(host_pattern='invalid_hsot'),
+                                    data_devices=DeviceSelection(paths=['']))
+            c = cephadm_module.create_osds(bad_dg)
+            out = wait(cephadm_module, c)
+            assert "Invalid 'host:device' spec: host not found in cluster" in out
+
```
Comment (adk3798) [inline src/pybind/mgr/cephadm/tests/test_cephadm.py:819]:
```
cephadm/tests/test_cephadm.py:821:5: E303 too many blank lines (2)
1     E303 too many blank lines (2)
```
flake8 error causing make check to fail

**Review by adk3798 — CHANGES_REQUESTED** [review]

File: src/pybind/mgr/cephadm/tests/test_cephadm.py (hunk 2 above)
Comment (rkachach) [reply src/pybind/mgr/cephadm/tests/test_cephadm.py:819]: Fixed :+1:

**Comment by adk3798** [comment]
http://pulpito.front.sepia.ceph.com/adking-2022-03-29_23:26:23-orch:cephadm-wip-adk-testing-2022-03-29-1402-distro-basic-smithi/

Failures:
https://tracker.ceph.com/issues/44587
https://tracker.ceph.com/issues/49287
1 machine failed to install kernel fast enough

**Pull request merged**